server/src/main/java/io/spine/server/entity/RecordBasedRepository.java

```diff
@@ -121,6 +128,90 @@ public void store(E entity) {
         return result;
     }
 
+    /**
+     * Applies a given {@link Migration} to an entity with the given ID.
+     *
+     * <p>The operation is performed in three steps:
+     * <ol>
+     *     <li>Load an entity by the given ID.
+     *     <li>Transform it through the migration operation.
+     *     <li>Store the entity back to the repository or delete it depending on the migration
+     *         configuration.
+     * </ol>
+     *
+     * <p>This operation is only supported for entities that are
+     * {@linkplain TransactionalEntity transactional}.
+     *
+     * @throws IllegalArgumentException
+     *         if the entity with the given ID is not found in the repository
+     * @throws IllegalStateException
+     *         if the repository manages a non-transactional entity type
+     *
+     * @see Migration
+     */
+    @SuppressWarnings("unchecked") // Checked at runtime.
+    @Experimental
+    public final <T extends TransactionalEntity<I, S, ?>>
+    void applyMigration(I id, Migration<I, T, S> migration) {
+        checkNotNull(id);
+        checkNotNull(migration);
+        checkEntityIsTransactional();
+
+        E entity = findOrThrow(id);
+        migration.applyTo((T) entity, (RecordBasedRepository<I, T, S>) this);
+        if (migration.physicallyRemoveRecord()) {
+            delete(id);
+            migration.finishCurrentOperation();
+        } else {
+            store(entity);
```

Comment:
It looks a bit odd to me that we `finish` the operation only if the migration removes a record. Why is it so?